central/credentialexpiry/service/service_impl_test.go

```diff
@@ -54,3 +66,125 @@ func TestEnsureTLSAndReturnAddr(t *testing.T) {
 }
+
+func TestGetScannerV4CertExpiry(t *testing.T) {
+	expiryCur := types.TimestampNow()
+	expiryOld, err := types.TimestampProto(time.Date(2000, 01, 01, 1, 1, 1, 1, time.UTC))
+	require.NoError(t, err)
+
+	successExpiryFunc := genGetExpiryFunc(map[mtls.Subject]*types.Timestamp{
+		mtls.ScannerV4IndexerSubject: expiryOld,
+		mtls.ScannerV4MatcherSubject: expiryCur,
+	})
+
+	matcherSuccessExpiryFunc := genGetExpiryFunc(map[mtls.Subject]*types.Timestamp{
+		mtls.ScannerV4MatcherSubject: expiryCur,
+	})
+
+	indexerSuccessExpiryFunc := genGetExpiryFunc(map[mtls.Subject]*types.Timestamp{
+		mtls.ScannerV4IndexerSubject: expiryOld,
+	})
+
+	errorExpiryFunc := genGetExpiryFunc(nil)
+
+	allScannerConfigs := map[mtls.Subject]*tls.Config{
+		mtls.ScannerV4IndexerSubject: {},
+		mtls.ScannerV4MatcherSubject: {},
+	}
+
+	noScannerConfigs := map[mtls.Subject]*tls.Config{}
+
+	// For readability.
+	feature := true
+	errorExpected := true
+	getIntegrationError := true
+	integrationExists := true
+
+	testCases := []struct {
+		desc                string
+		featureEnabled      bool
+		errorExpected       bool
+		scannerConfigs      map[mtls.Subject]*tls.Config
+		getIntegrationError bool
+		integrationExists   bool
+		expiryFunc          func(context.Context, mtls.Subject, *tls.Config, string) (*types.Timestamp, error)
+		expiryExpected      *types.Timestamp
+	}{
+		{
+			"error if feature disabled",
+			!feature, errorExpected, allScannerConfigs, !getIntegrationError, integrationExists, successExpiryFunc, expiryOld,
+		},
+		{
+			"error if scanner config missing",
+			feature, errorExpected, noScannerConfigs, !getIntegrationError, integrationExists, successExpiryFunc, expiryOld,
+		},
+		{
+			"error if failure getting scanner v4 integration",
+			feature, errorExpected, allScannerConfigs, getIntegrationError, integrationExists, successExpiryFunc, expiryOld,
+		},
+		{
+			"error if no scanner v4 integration",
+			feature, errorExpected, allScannerConfigs, !getIntegrationError, !integrationExists, successExpiryFunc, expiryOld,
+		},
+		{
+			"error if fail contacting both indexer and matcher",
+			feature, errorExpected, allScannerConfigs, !getIntegrationError, integrationExists, errorExpiryFunc, expiryOld,
+		},
+		{
+			"success if indexer and matcher able to be contacted",
+			feature, !errorExpected, allScannerConfigs, !getIntegrationError, integrationExists, successExpiryFunc, expiryOld,
+		},
+		{
+			"success if contacted matcher but failed contacting indexer",
+			feature, !errorExpected, allScannerConfigs, !getIntegrationError, integrationExists, matcherSuccessExpiryFunc, expiryCur,
+		},
+		{
+			"success if contacted indexer but failed contacting matcher",
+			feature, !errorExpected, allScannerConfigs, !getIntegrationError, integrationExists, indexerSuccessExpiryFunc, expiryOld,
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			if tc.featureEnabled {
+				testutils.MustUpdateFeature(t, features.ScannerV4, true)
```

Comment:
Question: Shouldn't you record the value of `features.ScannerV4` somewhere and restore it after running these test cases so that the feature flag updates are only local to these tests? Or is that happening already somewhere else or is being done implicitly somewhere?

Reply:
Shouldn't be needed - the implementation uses `t.Setenv()` to enable the feature, which isolates the change to the test.
https://github.com/stackrox/stackrox/blob/f96b83a20010fbdc098c6286745b832311c0072c/pkg/testutils/feature_flag.go#L19-L25

From the [t.Setenv docs](https://pkg.go.dev/testing#T.Setenv):
> Setenv calls os.Setenv(key, value) and uses Cleanup to restore the environment variable to its original value after the test.